Userland/Libraries/LibDSP/Resampler.h

```diff
@@ -0,0 +1,192 @@
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Array.h>
+#include <AK/Math.h>
+#include <AK/OwnPtr.h>
+#include <AK/Span.h>
+#include <AK/Vector.h>
+#include <AK/TypedTransfer.h>
+#include <LibC/math.h>
+#include <LibDSP/FIRFilter.h>
+#include <LibDSP/Window.h>
+
+namespace DSP {
+
+constexpr double normalized_sinc(double phi)
+{
+    if (AK::abs(phi) < NumericLimits<double>::epsilon()) {
+        return 1.0;
+    } else {
+        phi *= AK::Pi<double>;
+        return AK::sin(phi) / phi;
+    }
+}
+
+class InterpolatedSinc {
+public:
+    static ErrorOr<InterpolatedSinc> create(size_t sinc_taps, size_t oversample)
+    {
+        // Oversample is the number of lookup values between the taps. The more, the less aliasing noise.
+        // TODO: Use something better than linear interpolation so that we can save memory.
+
+        // +1 because we want the taps to be the same amount to the left and to the right.
+        // And another +1 because of an additional tap at (-TapCount - 1) to protect against bad lookups
+        // coming from rounding errors.
+        size_t const sinc_lookup_table_size = (2 * sinc_taps + 2) * oversample;
+
+        Vector<double> sinc_table;
+        TRY(sinc_table.try_resize(sinc_lookup_table_size));
+
+        for (ssize_t k = -sinc_taps - 1; k <= static_cast<ssize_t>(sinc_taps); k++) {
+            for (size_t i = 0; i < oversample; i++) {
+                double const sinc_index = k + static_cast<double>(i) / static_cast<double>(oversample);
+                size_t const window_index = (k + sinc_taps + 1) * oversample + i;
+                double const window = Window<double>::blackman_harris(window_index, sinc_lookup_table_size);
+                sinc_table[window_index] = normalized_sinc(sinc_index) * window;
+            }
+        }
+
+        return InterpolatedSinc(move(sinc_table), sinc_taps, oversample);
+    }
+
+    double operator()(double phi) const
+    {
+        double const fraction = phi - static_cast<i32>(phi);
+
+        // +1 to skip the rounding error buffer tap
+        size_t const n = static_cast<size_t>(phi * m_sinc_lookup_oversample) + m_index_offset;
+
+        double const y1 = m_sinc_table[n];
+        double const y2 = m_sinc_table[n + 1];
+
+        return y1 + fraction * (y2 - y1);
+    }
+
+private:
+    InterpolatedSinc(Vector<double>&& table, size_t taps, size_t oversample)
+        : m_sinc_table(table), m_sinc_taps(taps), m_sinc_lookup_oversample(oversample), m_index_offset((m_sinc_taps + 1) * m_sinc_lookup_oversample)
+    {
+    }
+
+    Vector<double> m_sinc_table;
+    size_t const m_sinc_taps;
+    size_t const m_sinc_lookup_oversample;
+    size_t const m_index_offset;
+};
+
+template<typename SampleType>
+class SincResampler {
+public:
+    SincResampler(double ratio, Function<double(double)>&& sinc_function, size_t sinc_taps, Vector<SampleType>&& input_buffer, FIRFilter<SampleType, double>&& lowpass)
```

Comment:
The amount of taps shouldn't be a parameter here, you should retrieve it from the passed sinc function. Then we can't use Function, but that's already kind of pointless; using a concept for something that has a `size_t taps() const` method will be useful:

```c++
template<typename T>
concept SincFunction = requires (T const function, double input) {
    { function(input) } -> SameAs<double>;
    { function.taps() } -> SameAs<size_t>;
};
```